### ISSUE TITLE
Question: parallel process() ?

### ISSUE THREAD

**Issue by ksloan** [jeffbski/redux-logic#32]
https://jsfiddle.net/jeffbski/954g5n7h/

Why is it that in the above fiddle, if you click the 'fetch users' button 5 times it takes 10s to respond? Shouldn't the next event either cancel the previous one, or at least be done in parallel? 

Even if you dispatch a 'USERS_FETCH_CANCEL' (by clicking 'cancel') between each 'USERS_FETCH' you still get this behaviour. The request is still in pending state, even after clicking cancel, and additional requests are waiting for previous ones.

Thanks!

**Comment by jeffbski** [comment]
@ksloan thanks for the question. 

## tl;dr

 - Browsers limit concurrent connections under http 1.x. 
 - redux-logic cancellation/latest cancels the use of the outdated responses
 - redux-logic can automatically perform a full xhr cancel at network level if you return RxJS Ajax observable
 - redux-logic can perform xhr cancel at network with axios with a few extra lines of code. 

## Details

The reason is that the browser is limiting the number of requests to the same server, typically 4-6 depending on the browser. When http2 is in use, those limits should disappear since they are pipelined in the same connection.

So while we are still in an http 1.x world the browser will limit us.

redux-logic does trigger a cancellation when `latest: true` and a concurrent request occurs or if a cancelType: actionType is set and received, however how far this cancellation goes depends on what you have implemented.

In all cases it will ignore the responses from the old or cancelled requests. So this prevents the actions from being dispatched.

If you use the RxJS async to make your xhr http request then it will fully cancel the xhr request (at the network level) automatically when redux-logic is cancelling or gets a later request. RxJS async returns an observable (which is cancellable) and if you in turn provide that to redux-logic it can cancel things automatically. https://jsfiddle.net/jeffbski/uLh2add5/

If you use a promise based api like axios or fetch then they don't have a built-in mechanism for xhr network cancellation from the promise. axios does provide a way to cancel manually, so we can have redux-logic call that with a little extra code. fetch doesn't provide cancellation at all yet.

So to have axios perform a true cancel of the network request with redux-logic (so the other requests can instantly start running without queuing) you would do something like this. See https://jsfiddle.net/jeffbski/w95q5udg/ for working example. We basically listen to cancelled$ and then we can perform an axios xrh cancel.

```js
const usersFetchLogic = createLogic({
  type: USERS_FETCH,
  cancelType: USERS_FETCH_CANCEL,
  latest: true, // take latest only

  // use axios injected as httpClient from configureStore logic deps
  // we also have access to getState and action in the first argument
  // but they were not needed for this particular code
  // cancelled$ is a built-in observable that indicates when cancel or take latest occurs
  process({ httpClient, cancelled$ }, dispatch, done) {
    var source = httpClient.CancelToken.source(); // axios.CancelToken
    cancelled$.subscribe(() => source.cancel('Request cancelled')); // xhr cancel
  
    // the delay query param adds arbitrary delay to the response
    httpClient.get(`https://reqres.in/api/users?delay=${delay}`, {
      cancelToken: source.token
    })      
    ... // same code as normal
  }
});
```

## Summary

 - http 1.x browsers limit concurrent connections to same server
 - http 2 limits should go away due to pipelining in a single connection
 - redux-logic cancellation/latest does cancel the use of outdated responses
 - redux-logic used with RxJS Ajax for fetching can automatically cancel xhr request at network level
 - redux-logic used with axios requires a few extra lines of code to enable xrh cancel at network level
 - xhr cancel at network is nice because it also saves bandwidth

**Comment by ksloan** [comment]
Thank you for the detailed answer!

**Comment by jeffbski** [comment]
You are welcome! I'll eventually add this as a linked example but I want to make sure it is documented sufficiently.
